web3/src/main/java/org/hiero/mirror/web3/evm/contracts/execution/traceability/AbstractOpcodeTracer.java

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package org.hiero.mirror.web3.evm.contracts.execution.traceability;
+
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.HederaSystemContract;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.tuweni.bytes.Bytes;
+import org.hiero.mirror.common.domain.contract.ContractAction;
+import org.hiero.mirror.web3.common.ContractCallContext;
+import org.hiero.mirror.web3.convert.BytesDecoder;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.precompile.PrecompiledContract;
+import org.springframework.util.CollectionUtils;
+
+public abstract class AbstractOpcodeTracer {
+
+    @Getter
+    @Setter
+    protected Map<Address, HederaSystemContract> systemContracts = new HashMap<>();
```

Comment:
After moving to child, ensure it is thread safe and final.

```suggestion
    @Getter
    protected final Map<Address, HederaSystemContract> systemContracts = new ConcurrentHashMap<>();
```

Reply:
Fixed